### ISSUE TITLE
Problems with malformed gtest json files

### ISSUE THREAD

**Issue by rorth** [llvm/llvm-project#56491]
Since the switch to gtest sharding in the testsuite, I'm constantly seeing a `lit` failure that breaks the testsuite.  The `ninja stage2-check-all` run ends with
```pytb
PASS: SanitizerCommon-Unit :: ./Sanitizer-x86_64-Test/17/91 (65005 of 65005)
Traceback (most recent call last):
  File "/var/llvm/dist-amd64-release-stage2-A/tools/clang/stage2-bins/./bin/llvm-lit", line 48, in <module>
    main(builtin_parameters)
  File "/vol/llvm/src/llvm-project/dist/llvm/utils/lit/lit/main.py", line 112, in main
    selected_tests, discovered_tests)
  File "/vol/llvm/src/llvm-project/dist/llvm/utils/lit/lit/formats/googletest.py", line 213, in post_process_shard_results
    testsuites = json.load(f)['testsuites']
  File "/usr/lib/python3.7/json/__init__.py", line 296, in load
    parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
  File "/usr/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.7/json/decoder.py", line 340, in decode
    raise JSONDecodeError("Extra data", s, end)
json.decoder.JSONDecodeError: Extra data: line 32 column 1 (char 598)
```
which unfortunately giving no indication whatsoever which json file is malformed.  I had to add debug code to `llvm/utils/lit/lit/formats/googletest.py` to find what went wrong: the malformed file is `/var/llvm/dist-amd64-release-stage2-A/tools/clang/stage2-bins/tools/polly/unittests/Isl/./IslTests-Polly-Unit-3150-1-20.json` and indeed the file has a superfluous trailing brace.

There are at least two issues here:

- The file shouldn't have ever been created this way in the first place.
- `lit` should give a proper indication of the wrong file, rather than just raising an exception.

I'm attaching the broken file (as .txt although it's .json actually).
[IslTests-Polly-Unit-3150-1-20.txt](https://github.com/llvm/llvm-project/files/9092081/IslTests-Polly-Unit-3150-1-20.txt)

**Comment by rorth** [comment]
FWIW I'm seeing exactly the same issue on Debian 11/sparc64 (`sparc64-unknown-linux-gnu`), so this is certainly not a Solaris issue.  Maybe SPARC, maybe big-endian?

**Comment by yuanfang-chen** [comment]
That is plausible. If it is deterministic, maybe file a ticket against gtest?

**Comment by rorth** [comment]
Unfortunately, this theory has just been falsified: while testing llvm-15.0.0-rc3 on `x86_64-pc-linux-gnu`, another `polly` `*.json` file proved equally malformed.  Neither SPARC nor Solaris in sight...

**Comment by rorth** [comment]
There was just another report of this issue on [LLVM 15.0.1 was tagged](https://discourse.llvm.org/t/llvm-15-0-1-was-tagged/65381/14) for `arm64-apple-darwin21.6.0`.  This is in no way restricted to a single platform, but getting more and more prevalent.

**Comment by efriedma-quic** [comment]
I saw what I think is the same issue on the reverse-iteration buildbot.  Didn't realize it was impacting release testing.

/cherry-pick 2c29268bfcc84c3b94bcb0aa34b7ef9c9bd9af01

**Comment by yuanfang-chen** [comment]
@efriedma-quic Thanks for the pointer and the cherry-picks.

I added a check in the LIT that could prevent this for unit tests and it would issue warnings if a gtest binary is added multiple times. 2380c1b609631c0

**Comment by efriedma-quic** [comment]
Let's keep this open until the fix is merged to 15.0.2

**Comment by songdongsheng** [comment]
I confirm this patch fixed the issue on SLE 15 SP4.
